server-adapters/resteasy-reactor-netty/src/main/java/org/jboss/resteasy/plugins/server/reactor/netty/ChunkOutputStream.java

```diff
@@ -1,156 +1,139 @@
-           SinkSubscriber.subscribe(completionSink, Mono.empty());
-       } else {
-           byteSink.emitComplete(EMIT_FAILURE_HANDLER);
-       }
-   }
-
-   @Override
-   public void write(byte[] bs, int off, int len) {
-       try {
-           asyncWrite(bs, off, len).get();
-       } catch (final InterruptedException ie) {
-           Thread.currentThread().interrupt();
-           throw new RuntimeException(ie);
-       } catch (final ExecutionException ee) {
-           throw new RuntimeException(ee);
-       }
-   }
-
-   @Override
-   public void flush() {
-       try {
-           asyncFlush().get();
-       } catch (final InterruptedException ie) {
-           Thread.currentThread().interrupt();
-           throw new RuntimeException(ie);
-       } catch (final ExecutionException ee) {
-           throw new RuntimeException(ee);
-       }
-   }
-
-   @Override
-   public CompletableFuture<Void> asyncFlush() {
-
-       // [AG] Discuss with @crankydillo.  Here is my understanding:
-       //   - asyncFlush is used mainly for SSE.
-       //   - The idea seems to be to send the element immediately without
-       //     waiting for the rest of the elements.  Anyway, that's what SSE is.
-       //   - But, at the same time, it is trying to not overload.  So, kind of
-       //     doing backpressure.  Only if the previous flush is complete,
-       //     then request the next element from the app.
-       //   - The backpressure mechanism is already built into Reactor Netty.  But,
-       //     the question is how to communicate that.
-       //   - Please see https://projectreactor.io/docs/netty/release/reference/index.html#_sse.
-       //     It reads `The flushing strategy is "flush after every element" emitted
-       //     by the provided Publisher`.  So, we do not need to call flush individually.
-       //
-       //   In summary, we are not controlling flushing, instead Reactor Netty does.
-       //   Also, this#asyncWrite will manage the backpressure.  I assume Reactor Netty
-       //   will backpressure if it cannot flush, and that would mean an element won't be
-       //   requested so asyncWrite won't complete until the element is pulled.  Also,
-       //   asyncFlush is called right after an asyncWrite.  So, asyncFlush looks
-       //   useless.
-
-       return CompletableFuture.completedFuture(null);
-   }
-
-   @Override
-   public CompletableFuture<Void> asyncWrite(final byte[] bs, int offset, int length) {
-        final CompletableFuture<Void> cf = new CompletableFuture<>();
-        if (!started) {
-            byteSink = byteSinkSupplier.get();
-            parentResponse.committed();
-            started = true;
+public class ChunkOutputStream extends AsyncOutputStream {
+
+    /**
+     * This is the {@link Mono} that we return from {@link ReactorNettyJaxrsServer.Handler#handle(HttpServerRequest,
+     * HttpServerResponse)}
+     */
+    protected final Sinks.Empty<Void> completionSink;
```

Comment:
Is there a reason this can't remain private?

Reply:
There is no reason to keep this private. I will update this.